flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java

```diff
@@ -156,14 +177,29 @@ public InputStatus pollNext(ReaderOutput<T> output) throws Exception {
 
         // we need to loop here, because we may have to go across splits
         while (true) {
+            // Check if the previous record count reached the limit of rateLimiter.
+            if (rateLimitPermissionFuture != null && !rateLimitPermissionFuture.isDone()) {
```

Comment:
In current implementation, we need to check here for **each record `** even rateLimitPermissionFuture is null , I believe this will introduce a performance regression. We can avoid this regression as we've known that `rateLimiterStrategy` nullability in constructor.